### ISSUE TITLE
Role execution in loop  (using `include_role`) may happen in wrong order

### ISSUE THREAD

**Issue by Cougar** [ansible/ansible#54173]
##### SUMMARY
Role execution order may change when executing roles for different hosts but having exactly the same input variables.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
Core

##### ANSIBLE VERSION
```
ansible 2.7.9
  config file = /home/cougar/.ansible.cfg
  configured module search path = ['/home/cougar/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/cougar/test/ansible-bug-uuid/.venv/lib64/python3.7/site-packages/ansible
  executable location = /home/cougar/test/ansible-bug-uuid/.venv/bin/ansible
  python version = 3.7.2 (default, Dec 30 2018, 16:18:15) [GCC]
```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT
openSUSE Leap 42.3, Python 3.7.2, Ansible in venv.

##### STEPS TO REPRODUCE
This is the directory layout:
```
.
├── inventory
│   └── hosts
├── roles
│   └── test
│       └── tasks
│           └── main.yml
└── test.yml
```

`roles/test/tasks/main.yml`:
```yaml
---

- name: show task items
  debug:
    msg: "run item {{ item.name }}"
```

`inventory/hosts`:
```
host1
host2
```

`test.yml` playbook:
```yaml
#!/usr/bin/env ansible-playbook

- hosts: host1, host2
  gather_facts: no

  vars:
    tasklist:
      - name: task1
        x: "{{ inventory_hostname }}"
      - name: task2

  tasks:
    - include_role:
        name: test
      loop: "{{ tasklist }}"
```

Run it like this
```sh
python3 -mvenv .venv
. .venv/bin/activate
pip install ansible
./test.yml -i inventory/
```

Look at "run item ?" order. It will be:
- host1 task1
- host1 task2
- **host2 task2 !!!!!**
- host2 task1

This happens because of the vars list where second member "task2" is exactly the same for all hosts.

```yaml
  vars:
    tasklist:
      - name: task1
        x: "{{ inventory_hostname }}"
      - name: task2
```

My guess is that Ansible executes roles in order but using task uuid. uuid will be the same for tasks which have exactly the same configuration (variables) and our "task2" is exactly the same for all hosts.

Now when Ansible runs tasks using uuid, it matches all hosts and executes these in parallel. It is good optimization but as long all previous tasks are not executed in some hosts, it should not do that.

Everything works fine when adding something "unique" to both list entries like this:
```yaml
  vars:
    tasklist:
      - name: task1
        x: "{{ inventory_hostname }}"
      - name: task2
        y: "{{ inventory_hostname }}"
```

At the same normal tasks in playbook work fine. There are no problems with this playbook task without using role:

```yaml
    - name: show task items
      debug:
        msg: "run item {{ item.name }}"
      loop: "{{ tasklist }}"
```

##### EXPECTED RESULTS

```
PLAY [host1, host2] ********************************************************************

TASK [include_role : test] *************************************************************

TASK [test : show task items] **********************************************************
ok: [host1] => {
    "msg": "run item task1"
}

TASK [test : show task items] **********************************************************
ok: [host1] => {
    "msg": "run item task2"
}

TASK [test : show task items] **********************************************************
ok: [host2] => {
    "msg": "run item task1"
}

TASK [test : show task items] **********************************************************
ok: [host2] => {
    "msg": "run item task2"
}

PLAY RECAP *****************************************************************************
host1                      : ok=2    changed=0    unreachable=0    failed=0   
host2                      : ok=2    changed=0    unreachable=0    failed=0   
```

##### ACTUAL RESULTS

```paste below
PLAY [host1, host2] ********************************************************************

TASK [include_role : test] *************************************************************

TASK [test : show task items] **********************************************************
ok: [host1] => {
    "msg": "run item task1"
}

TASK [test : show task items] **********************************************************
ok: [host1] => {
    "msg": "run item task2"
}
ok: [host2] => {
    "msg": "run item task2"
}

TASK [test : show task items] **********************************************************
ok: [host2] => {
    "msg": "run item task1"
}

PLAY RECAP *****************************************************************************
host1                      : ok=2    changed=0    unreachable=0    failed=0   
host2                      : ok=2    changed=0    unreachable=0    failed=0   
```

**Comment by sivel** [comment]
Duplicate of https://github.com/ansible/ansible/issues/36978
